ouroboros-network/src/Ouroboros/Network/Protocol/VersionNegotiation/Type.hs

```diff
@@ -0,0 +1,424 @@
+ <> mconcat [    CBOR.encode (protocolVersion v)
+              <> encodeVersionData v
+            | v <- vs
+            ]
+
+-- |
+-- decode a list of @'ProtocolVersion'@. Fail when a the received map has
+-- unsorted keys (which also means that it will fail if keys are duplicated).
+--
+decodeKnownVersions :: forall ver r s.
+                       ( Ord ver
+                       , Enum ver
+                       , Show ver
+                       , Serialise ver
+                       )
+                    => KnownVersions ver r
+                    -> CBOR.Decoder s (KnownVersions ver r)
+decodeKnownVersions knownVersions = CBOR.decodeMapLen >>= go Nothing []
+  where
+    go !ver0 !rs !l | l <= 0 = return $ KnownVersions $ reverse rs
+                    | otherwise = do
+      ver <- CBOR.decode
+      if Just ver <= ver0
+        then fail "decodeVersionMap: unsorted map"
+        else case findProtocolVersion ver knownVersions of
+          Nothing -> CBOR.decodeTerm >> go (Just ver) rs (pred l)
+          Just pv -> decodeVersionData pv >>= \r -> go (Just ver) (r : rs) (pred l)
+
+-- |
+-- Smart constructor for the sorted list of protocol versions.
+--
+newKnownVersions :: Ord ver => [ProtocolVersion ver r] -> KnownVersions ver r
+newKnownVersions ps = KnownVersions $ sortBy (\(ProtocolVersion verA _) (ProtocolVersion verB _) -> compare verA verB) ps
+
+-- |
+-- Find protocol version in a list of known versions.  It assumes that the
+-- @knownVersion@ list is sorted with respect to @ver@ (which is fullfiled by
+-- @newKnownVersions@ smart constructor).
+--
+findProtocolVersion
+  :: Ord ver
+  => ver
+  -> KnownVersions ver r
+  -> Maybe (ProtocolVersion ver r)
+findProtocolVersion ver (KnownVersions knownVersions) = go (length knownVersions) knownVersions
+    where
+      go _   []   = Nothing
+      go _   [pv] | protocolVersion pv == ver = Just pv
+                  | otherwise                 = Nothing
+      go len pvs  = let l = len `div` 2
+                    in case splitAt l pvs of
+                      (pvs', pv : pvs'') | protocolVersion pv == ver -> Just pv
+                                         | protocolVersion pv >  ver -> go l pvs'
+                                         | protocolVersion pv <  ver -> go (l - 1) pvs''
+                      _                  -> error "findProtocolVersion error"
+
+data ProtocolVersionDescription m where
+       ProtocolVersionDescription :: (Bounded ptcl, Enum ptcl)
+                                  => ptcl
+                                  -> MiniProtocolDescription ptcl m
+                                  -> ProtocolVersionDescription m
+
+-- |
+-- State transitions of the version negotation protocol.
+--
+data VersionNegotiationProtocol ver r where
+    StPropose :: VersionNegotiationProtocol ver r
+    StConfirm :: VersionNegotiationProtocol ver r
+    StDone    :: VersionNegotiationProtocol ver r
+
+
+instance Protocol (VersionNegotiationProtocol ver r) where
```

Comment:
What does sending _every_ version and its data up-front buy us?

In the normal case, the server knows the client version, so it would be better to just send one version number and its data. The server will accept it; we can move on. By sending all versions, we're optimizing for the comparatively rare case in which the client is running a newer version than that of the server. Instead, why not just accept that in this case we'll have to do more than one round trip? The server can counter-offer, the client can counter-counter-offer, etc. etc.

This way, we get size-limits built-in assuming each version data blob has a natural bounded size. We wouldn't have to put down an arbitrary limit. Assuming the set of versions known to a peer is finite, and the version data for each peer has a bounded encoded size, the version negotiation protocol is therefore bounded in time and space use: the peer will eventually agree or give up once it has tried all of its finitely-many versions, and it retains at most the order of the size of the biggest possible encoded version datum.

Reply:
The intention was never to send *every* version supported by the software but a list of versions supported by the current configuration. I'd expect that to correspond to two versions, latest version implemented and the previous version to facilitate smooth upgrades.

Being able to do version negotiation in one RTT is a huge advantage . The transmission time for one version is expected to be the same for one version and multiple versions as long as the overall message size is below 10 TCP segments so presenting them all at once is basically free.